### PR TITLE
Add example app

### PR DESCRIPTION
**Pull request by Kijewski**
I thought: We explain a lot in the book, use all functions in the tests, but we have no real example application so you can see rinja in action in a half-way realistic setting.

Todo:
- [x] write a README
- [x] add some comments in the source
- [x] adding a french translation in the templates

If that is a good example, I don't know. I was too lazy to write the ubiquitous TODO-app example.

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
It's a very good idea! And basically, any example is better than no examples. :wink:

File: examples/actix-web-app/templates/greet.html [hunk 1]
```diff
@@ -0,0 +1,44 @@
+{% extends "_layout.html" %}
+
+{% block title -%}
+    {%- match lang -%}
+        {%- when Lang::en -%} Hello, {{name}}!
+        {%- when Lang::de -%} Hallo, {{name}}!
```
Comment (GuillaumeGomez) [inline examples/actix-web-app/templates/greet.html:6]: For french part :D

```jinja2
{%- when Lang::fr -%} Bonjour, {{name}}!
```

File: examples/actix-web-app/templates/greet.html [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    {%- endmatch -%}
+{%- endblock %}
+
+{%- block content -%}
+    <h1>
+        {%- match lang -%}
+            {%- when Lang::en -%} Hello
+            {%- when Lang::de -%} Hallo
```
Comment (GuillaumeGomez) [inline examples/actix-web-app/templates/greet.html:14]: For french part :D

```jinja2
{%- when Lang::fr -%} Bonjour!
```

File: examples/actix-web-app/templates/greet.html [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        {%- endmatch -%}
+    </h1>
+    <p>
+        {%- match lang -%}
+            {%- when Lang::en -%}
+                Hello, <strong>{{name}}</strong>, nice to meet you! {#-~#}
+                I'm a <a href="https://rinja.readthedocs.io/en/latest/">Rinja</a> example application.
+            {%- when Lang::de -%}
```
Comment (GuillaumeGomez) [inline examples/actix-web-app/templates/greet.html:22]: For french part :D

```jinja2
{%- when Lang::en -%}
                Bonjour, <strong>{{name}}</strong>, ravi de vous rencontrer ! {#-~#}
                Je suis une application d'exemple de <a href="https://rinja.readthedocs.io/en/latest/">Rinja</a>.
```

File: examples/actix-web-app/templates/greet.html [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+                Hallo, <strong>{{name}}</strong>, schön dich kennenzulernen! {#-~#}
+                Ich bin eine <a href="https://rinja.readthedocs.io/en/latest/">Rinja</a>-Beispielanwendung.
+        {%- endmatch -%}
+    </p>
+    <h2>
+        <a href="{{ req.url_for("index_handler", [lang])? }}?name={{ name|urlencode }}">
+            {%- match lang -%}
+                {%- when Lang::en -%} Back to the first page.
+                {%- when Lang::de -%} Zurück zur ersten Seite.
```
Comment (GuillaumeGomez) [inline examples/actix-web-app/templates/greet.html:31]: {%- when Lang::en -%} Retour à la première page.

File: examples/actix-web-app/templates/greet.html [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        </a>
+    </h2>
+
+    <ul id="lang-select">
+        {%- if lang != Lang::en -%}
+            <li><a href="{{ req.url_for("greeting_handler", [Lang::en])? }}?name={{ name|urlencode }}">This page in English</a></li>
+        {%- endif -%}
+        {%- if lang != Lang::de -%}
+            <li><a href="{{ req.url_for("greeting_handler", [Lang::de])? }}?name={{ name|urlencode }}">Diese Seite auf deutsch.</a></li>
+        {%- endif -%}
```
Comment (GuillaumeGomez) [inline examples/actix-web-app/templates/greet.html:42]: For french part :D

```jinja2
{%- if lang != Lang::fr -%}
            <li><a href="{{ req.url_for("greeting_handler", [Lang::de])? }}?name={{ name|urlencode }}">Cette page est en français.</a></li>
        {%- endif -%}
```

File: examples/actix-web-app/templates/greet.html [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        <a href="{{ req.url_for("index_handler", [lang])? }}?name={{ name|urlencode }}">
+            {%- match lang -%}
+                {%- when Lang::en -%} Back to the first page.
+                {%- when Lang::de -%} Zurück zur ersten Seite.
+            {%- endmatch -%}
+        </a>
+    </h2>
+
+    <ul id="lang-select">
+        {%- if lang != Lang::en -%}
```
Comment (GuillaumeGomez) [inline examples/actix-web-app/templates/greet.html:37]: One question: shouldn't it be `==` and also `else if` following?

File: examples/actix-web-app/templates/index.html [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        {%- when Lang::en -%} Hello
+        {%- when Lang::de -%} Hallo
+    {%- endmatch -%}
+{%- endblock %}
+
+{%- block content -%}
+    <h1>
+        {%- match lang -%}
+            {%- when Lang::en -%} Hello
+            {%- when Lang::de -%} Hallo
```
Comment (GuillaumeGomez) [inline examples/actix-web-app/templates/index.html:14]: For french part :D

```jinja2
{%- when Lang::fr -%} Bonjour
```

File: examples/actix-web-app/templates/index.html [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    >
+        <p>
+            {%- match lang -%}
+                {%- when Lang::en -%}
+                    I would like to say <em>hello</em>. {#-~#}
+                    Would you please tell me your name?
+                {%- when Lang::de -%}
+                    Ich möchte dir gerne <em>hallo</em> sagen. {#-~#}
+                    Bitte nenne mir doch deinen Namen!
+            {%- endmatch -%}
```
Comment (GuillaumeGomez) [inline examples/actix-web-app/templates/index.html:30]: For french part :D

```jinja2
{%- when Lang::fr -%} Je voudrais vous dire <em>bonjour</em>. {#-~#}
Pourriez-vous me donner votre nom ?
```

File: examples/actix-web-app/templates/index.html [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        </p>
+        <p>
+            <label>
+                {%- match lang -%}
+                    {%- when Lang::en -%} My name is
+                    {%- when Lang::de -%} Ich heiße
```
Comment (GuillaumeGomez) [inline examples/actix-web-app/templates/index.html:36]: For french part :D

```jinja2
{%- when Lang::fr -%} Je m'appelle
```

File: examples/actix-web-app/templates/index.html [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+                    style="width: 10em"
+                />
+            </label>
+        </p>
+        <p>
+            <label>
+                <button type="submit">
+                    {%- match lang -%}
+                        {%- when Lang::en -%} Greet me, then!
+                        {%- when Lang::de -%} Dann begrüße mich!
```
Comment (GuillaumeGomez) [inline examples/actix-web-app/templates/index.html:52]: For french part :D

```jinja2
{%- when Lang::fr -%} Saluons-nous !
```

File: examples/actix-web-app/templates/index.html [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        </p>
+    </form>
+
+    <ul id="lang-select">
+        {%- if lang != Lang::en -%}
+            <li><a href="{{ req.url_for("index_handler", [Lang::en])? }}">This page in English</a></li>
+        {%- endif -%}
+        {%- if lang != Lang::de -%}
+            <li><a href="{{ req.url_for("index_handler", [Lang::de])? }}">Diese Seite auf deutsch.</a></li>
+        {%- endif -%}
```
Comment (GuillaumeGomez) [inline examples/actix-web-app/templates/index.html:65]: For french part :D

```jinja2
{%- when Lang::fr -%}
    <li><a href="{{ req.url_for("index_handler", [Lang::en])? }}">Cette page est en français.</a></li>
```

File: examples/actix-web-app/templates/index.html [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+                        {%- when Lang::de -%} Dann begrüße mich!
+                    {%- endmatch -%}
+                </button>
+            </label>
+        </p>
+    </form>
+
+    <ul id="lang-select">
+        {%- if lang != Lang::en -%}
+            <li><a href="{{ req.url_for("index_handler", [Lang::en])? }}">This page in English</a></li>
```
Comment (GuillaumeGomez) [inline examples/actix-web-app/templates/index.html:61]:
```suggestion
            <li><a href="{{ req.url_for("index_handler", [Lang::en])? }}">This page is in English.</a></li>
```

**Comment by Kijewski** [comment]
Thank you for the translations! :)

In the bottom line I wanted to list links to all languages that are currently not selected, that's why I used `if !=` without an `else`:

](https://i.imgur.com/k25aK2V.png)

**Comment by GuillaumeGomez** [comment]
Why not using an array then?

**Comment by Kijewski** [comment]
> Why not using an array then?

Sorry, I don't follow. What array?

**Comment by GuillaumeGomez** [comment]
Nevermind. However, there is duplicated code between `index.html` and `greet.html`. Why not putting them into a "footer.html" file or equivalent and `include` it?

**Comment by Kijewski** [comment]
I made the language selection footer a macro, to show that we have macros and how to use them.  An `include` would possibly be more conventional, but I thought the might be a good case to show another feature. What do you think?

**Comment by GuillaumeGomez** [comment]
I think it's a good idea. Well done!

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by Kijewski** [comment]
Thanks! And thank you for your reviews! :)

**Pull request merged**